### PR TITLE
GEODE-7018: Enable hostname validation for SSL over NIO

### PR DESCRIPTION
**Pull request by sboorlagadda**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/integrationTest/java/org/apache/geode/internal/net/SSLSocketHostNameVerificationIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        SocketCreator sc = SocketCreatorFactory.getSocketCreatorForComponent(CLUSTER);
+        engine =
+            sc.handshakeSSLSocketChannel(socket.getChannel(),
+                sc.createSSLEngine(this.localHost.getHostName(), 1234),
+                timeoutMillis,
+                false,
+                ByteBuffer.allocate(500),
+                new BufferPool(mock(DMStats.class)));
+      } catch (Throwable throwable) {
+        throwable.printStackTrace(System.out);
```
Comment (bschuchardt) [inline geode-core/src/integrationTest/java/org/apache/geode/internal/net/SSLSocketHostNameVerificationIntegrationTest.java]: No need to print this stack trace

File: geode-core/src/main/java/org/apache/geode/internal/net/SocketCreator.java [hunk 2]
```diff
@@ -933,6 +933,24 @@ public NioSslEngine handshakeSSLSocketChannel(SocketChannel socketChannel, SSLEn
       BufferPool bufferPool)
       throws IOException {
     engine.setUseClientMode(clientSocket);
+    if (!clientSocket) {
+      engine.setNeedClientAuth(sslConfig.isRequireAuth());
+    }
+
+    if (clientSocket) {
+      if (sslConfig.doEndpointIdentification()) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/net/SocketCreator.java]: this whole "if/else" statement is also in this class in configureClientSSLSocket. Should you extract it into a private method so both places use the same code?

Comment (sboorlagadda) [reply geode-core/src/main/java/org/apache/geode/internal/net/SocketCreator.java]: Thanks @dschneider-pivotal I have refactored it to a method. Both `SSLSocket` and `SSLEngine` returns a copy of SSLParameters to I could not completely do it inside the method. This new method returns params which are re-set into socket or engine.

File: geode-core/src/integrationTest/java/org/apache/geode/internal/net/SSLSocketHostNameVerificationIntegrationTest.java (hunk 1 above)
Comment (sboorlagadda) [reply geode-core/src/integrationTest/java/org/apache/geode/internal/net/SSLSocketHostNameVerificationIntegrationTest.java]: Thanks @bschuchardt. Removed the unwanted output

**Pull request merged**